R/merge_sce_list.R

```diff
@@ -445,3 +398,125 @@ get_altexp_attributes <- function(sce_list) {
   }
   return(altexp_attributes)
 }
+
+
+
+
+
+#' Prepare an updated list of metadata for the merged SCE
+#'
+#' The metadata will be reformatted to contain the following fields:
+#' - `library_id`:       vector of all library ids in the merged object
+#' - `sample_id`:        vector of all sample ids in the merged object
+#' - `library_metadata`: pre-merge metadata list for each library, named by SCE name
+#' - `sample_metadata`:  a data frame with all distinct rows from the pre-merge sample_metadata data frames,
+#'   with all values coerced to character
+#'
+#' @param metadata_list List of metadata to update
+#'
+#' @return Updated metadata list to include in the merged SCE
+prepare_merged_metadata <- function(metadata_list) {
+  # Define vectors of library and sample ids
+  metadata_library_ids <- metadata_list |>
+    purrr::map_chr(
+      purrr::pluck("library_id")
+    )
+
+  metadata_sample_ids <- metadata_list |>
+    purrr::map_chr(
+      purrr::pluck("sample_id")
+    )
+
+
+  # Grab names to check contents
+  transposed_names <- metadata_list |>
+    purrr::list_transpose() |>
+    names()
+
+  # first check that this library hasn't already been merged
+  if ("library_metadata" %in% transposed_names) {
+    stop(paste(
+      "This SCE object appears to be a merged object",
+      "We do not support merging objects with objects that have already been merged."
+    ))
+  }
+
+  # Create new version of metadata_list without any sample_metadata fields
+  # Note that this will work even if `sample_metadata` is not present
+  library_metadata <- metadata_list |>
+    purrr::map(
+      \(meta) purrr::assign_in(meta, "sample_metadata", purrr::zap())
```

Comment:
⚡

Reply:
I thought I commented about this, but maybe I didn't. what you have here obviously works, but I feel like it is a bit... over-engineered? I'd do a more direct manipulation:

```suggestion
      \(meta) meta[which(names(meta) != "sample_metadata")]
```
(using `which` to be safe)

Reply:
> over-engineered

What a nice euphemism for me! 😂